feat v0

moreno-moises opened the pull request:
# Description
This pull request introduces the initial implementation of `ofxGuiJson`, a module designed to manage GUI elements dynamically based on a JSON configuration. It integrates this functionality within an audio application to control various sound settings and an oscilloscope setup. The primary features include loading configurations, creating GUI elements, and binding actions based on the JSON configuration.

# Requirements
- Ensure OpenFrameworks is set up. Follow the [setup guide](https://openframeworks.cc/setup/).
   - I used [OF0.11.2](https://github.com/openframeworks/openFrameworks/releases/tag/0.11.2) 
- Ensure Visual Studio 2022 is set up for compiling.
- Required libraries that are included in the project:
  - [`ofxGui`](https://github.com/openframeworks/openFrameworks/tree/master/addons/ofxGui)
  - [`ofxMaxim`](https://github.com/micknoise/Maximilian)
- Ensure to move all addons and ofxGuiJson to OF's addons directory

# Issues Referenced
- [Trouble Configuring `nlohmann::json` Library](https://github.com/moreno-moises/ofxGuiJson/issues/2)
- [Dynamic Toggles Feature Incompletion](https://github.com/moreno-moises/ofxGuiJson/issues/3)


# Documentation update
- None

# Notes for Reviewer
- Action binding is currently implemented only for toggle elements. However, the system is designed to be scalable so that other elements can also have action bindings in the future.
- Instead of directly referencing specific GUI elements for binding actions as initially considered (e.g., using `getGuiElement()` in `ofApp` to fetch an element and bind actions directly), I implemented a different solution where actions are defined and bound within the `ofxGuiJson` class itself. 

# List of new distributable files (For software PR only)
- None

# Testing
#### example_audio
- [x] Successfully built the audio application on Windows.
- [x]  Verified that the audio controls (mute, frequency adjustment) responded as expected without any issues.
- [x] No crashes or unhandled exceptions occurred during the testing phase.
#### example_oscilloscope
- [x] Successfully built the oscilloscope application on Windows.
- [x] Ensured that all GUI elements were responsive and positioned accurately.
- [x] No crashes or unhandled exceptions occurred during the testing phase.
#### JSON Exception Handling
- [x] **File Absence**: Properly handled the absence of the JSON configuration file. Log output:
  - `[ error ] Failed to open JSON config file: C:\path\bin\data\config.json`
  - `[ error ] JSON config is null, cannot parse GUI elements.`
  - `[ error ] Configuration is null after loading.`
- [x] **Syntax Errors**: Correctly identified and reported syntax errors within the JSON file. Log output:
  - `[ error ] JSON parse error at byte 428: [json.exception.parse_error.101] parse error at line 20, column 16: syntax error while parsing object - unexpected string literal; expected '}'`

## Feature Tests
- None

# Shared files
- None

# Checklist to allow merge
- [ ] All dependent repositories used were on branch `master`
- Software
  - [ ] Get approval from the reviewer
  - [ ] Passed testing on Windows
  - [ ] Passed testing on macOS *(for major changes/GUI changes/ PRs adding files distributed with the EmotiBit software)*
  - [ ] Passed testing on linux (ubuntu) *(for major changes/GUI changes/ PRs adding files distributed with the EmotiBit software)*
  - [ ] Update software bundle version in `ofxEmotiBitVersion.h`
- [ ] doxygen style comments included for new code snippets
- [ ] Required documentation updated


## Screenshots:

moreno-moises committed:
added src files and examples

first commit in feat-v0 branch, adding all src files and initial two examples for ofxGuiJson

moreno-moises committed:
add: bin/data folder + JSON

nitin710 commented:
# Review 01
### About setup
  - [ ] It'll be nice to mention which OF version you used to test. It'll also be nice to mention the Visual studio version. I assume you used OF0.12+VS2022. *I got it to compile on OF0.11.2+VS2022*
  - [ ] Maybe the above information can go in the "requirements" section, but use your judgement.
  - [ ] I cloned [this link](https://github.com/micknoise/Maximilian) that is listed in the requirements section, but there was no mention of the additional step of moving the `ofxMaxim` from `Maximilian/cpp/openFrameworks` to `OF_ROOT/ofxMaxim`. That is what i did to be able to compile from source, but lemme me know if that was not correct. Either way, I think itll be good to mention any addional steps to do after cloning Maximillian to get the source to compile.
    - The error i got when trying to compile without moving the ofxMaxin to `OF_ROOT/Addons`
    - ![image](https://github.com/user-attachments/assets/2302cc1f-5fde-484a-bba4-455d980b9bdf)

### Examples
#### example_audio
- ✔️ compiled from source
- ✔️ Tested working
  - Frequency modulation works associated with the mouse movement
#### example_oscilloscope
- ✔️ compiled from source
- ✔️ tested working
  - toggle boxes show/hide check boxes.
  - [ ] Should we have sample functions linked to the toggle boxes that print something on the console if clicked? I was thinking that might be a more complete example of the Oscilloscope. i.e., "click here and this happens", BUT, if you think that complicates the update function, maybe its not worth it. What do you think?
  - [ ] I think it would be better if the example opened a window that was the size of the GUI. I mean, window_width = GUI_panel width. Currently, the window is smaller so it cuts out the gui panels.

## General comments
- [ ] I think it'll add to the clarity if you could add a "how to use this addon", probably in the addons `README.md`? Clearly stating
  - A config file is used to specify GUI elements
  - How to map functions to GUI elements, etc
  - My thought is that, if people can read a block diagram and understand the addone functionality, it may be used in the way it is intended to.
- [ ] Add [doxygen style comments](https://www.doxygen.nl/manual/docblocks.html) to the `ofxGuiJson.h` file. See [this](https://github.com/EmotiBit/ofxEmotiBit/blob/master/src/EmotiBitLsl.h) for example.
- [ ] Should the json file name in the example projects be changed from `config.json` to `guiConfig.json`? I think it explains better the use in reference to the example. Let me know what you think

@moreno-moises 

ToDo: Code review.

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+
+
+std::vector<std::shared_ptr<ofxPanel>>& ofxGuiJson::getPanels() {
+    return panels;
```
The usage suggests `panels` is returned to ofApp, which then calls `panel->draw();`. If `panels` is being sed outside of ofxGuiJson, should it still be private?

moreno-moises replied:
you are correct in suggesting that `panels` is treated as a public pointer, but is declared private. i moved the definition so it is now declared as public

nitin710 commented on src/ofxGuiJson.h:
```diff
@@ -9,49 +9,38 @@
 
 using json = nlohmann::json;
```
`Using` can be tricky here.
This implicitly means that any cpp file including `ofxGuiJson.h` now has to conform to your usage of json. I can see this causing potential bugs. Is there a reason why we just don't use `nlohmann::json` instead of [changing the namespace in all cpp files that include this header](https://stackoverflow.com/a/223047/17719756)?

nitin710 replied:
Also, any reason why we are not using `ofxJSON`?

moreno-moises replied:
the concern about using a namespace alias like using json = nlohmann::json; in a header file (ofxGuiJson.h) is indeed valid, especially in a shared or larger codebase. I am having trouble working with additional libraries and cannot resolve this error given the time. i will add a note to this in the source code for future development

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+void ofxGuiJson::setup(const std::string& configFile) {
+    loadJson(configFile);
+    setBackgroundColor();  // Set background color based on JSON config
+    parseGuiElements();
```
We should also use the return value to catch an exception, for example, a null JSON and stop the application. I don't know what happens now, if the config file is not present, since we we don't quit the application if parse fails.

moreno-moises replied:
exception handling is included in the loadJson function. proper logging is implemented to ensure that the user has clarity when JSON loading is errored

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+void ofxGuiJson::parseGuiElements() {
```
Should the return type be `int` to let the calling function know of a parse fail? Alternatively, you can exit the application from parse after logging appropriate message on the console.

moreno-moises replied:
this would require the caller to handle the error. If the caller neglects to check the return value, the error might go unnoticed, potentially leading to further issues down the execution path. logging is already implemented in this function in the event of a failed parse. i don't think this is necessary

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+        ofLogError() << "Invalid 'gui' configuration for panel.";
+        return;
+    }
+    auto& guiConfig = panelConfig["gui"];
```
Any reason for declaring `guiConfig` as a reference?

moreno-moises replied:
by declaring guiConfig as a reference, you avoid copying the JSON object. JSON objects can be large or contain nested structures, so copying them unnecessarily can lead to performance overhead. a reference provides direct access to the original JSON object within panelConfig, ensuring that any changes or accesses are performed directly on the data structure stored in memory, without the need for intermediate copies.

moreno-moises committed:
add: README.md

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+void ofxGuiJson::createAndAddToggle(const std::string& label, const std::string& baseId) {
```
is dynamically adding toggles functional? I tried it in the example_audio, but it did not generate any toggles on screen. However, when i enabled `logToConsole` I did see verbose log prints on the console, but I did not see the gui update. Just wondering if i used it wrong.
![image](https://github.com/user-attachments/assets/50fe4c52-b0fb-40bf-9336-c33f4b60fce0)

nitin710 replied:
I also tried pressing `b` to rebuild the GUI, but nothing happened on-screen. I added this section in the code to access rebuild().
```
    if (key == 'b')
    {
        guiManager.rebuildGui();
    }
```

nitin710 commented on src/ofxGuiJson.h:
```diff
@@ -9,49 +9,38 @@
-    std::vector<std::string> orderedElements;
 
     void parseGuiElements();
+    std::vector<std::shared_ptr<ofxPanel>>& getPanels();  // Return a reference to the vector of shared_ptr
```
This function returns a private variable, which defeats the purpose of calling something private. Are we sure this is the correct implementation?

moreno-moises replied:
this variable has since been made public. you were correct in pointing this out

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+    parseGuiElements();
+}
+
+void ofxGuiJson::loadJson(const std::string& relativePath) {
```
I think there is some confusion with passed variable `relativePath`.
In setup, it looks like a "filename" is passed. But then setup simply calls `loadJson` and passes that name, which now becomes relative path?

Looks like the variable name in setup should cahnge from configFile to `path` and `ofToDataPath` should be moved to setup?

Let me know if I;m wrong in understanding this.

moreno-moises replied:
you're correct in noting the confusion around the variable name and its usage in the setup and loadJson functions of the ofxGuiJson class. the naming should clearly reflect the purpose and flow of data through these functions. this section of the code has been updated according to your suggestions

nitin710 commented on src/ofxGuiJson.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+void ofxGuiJson::setBackgroundColor() {
```
Should setting the bgColor be a part of this class?
I just ask because all we do is call `ofSetBackgroundColor`. It has nothing else to do with this class, except being able to read the color from a file. that is nice, but then it locks users to specifying bg color in the config file.

moreno-moises replied:
i think it is appropriate for this to be a part of the class as it aligns with the current functionality of allowing the user to customize the GUI through a JSON

nitin710 commented on example_audio/src/ofApp.cpp:
```diff
@@ -1,11 +1,12 @@
 #include "ofApp.h"
 
 void ofApp::setup() {
```
Should there be a `ofLogToConsole` call here so that we can see errors/warnings/info be printed on the console?

moreno-moises replied:
i added the appropriate lines of code in the setup functions so that we can see errors/warnings/info printed on the console

nitin710 reviewed (changes requested):
I think overall it looks good.
Some questions about 
1. Public/private scope
2. exception handling (since this is very user facing)
3. console logging
4. A high level block diagram(already discussed)

moreno-moises committed:
add: readme project structure flow diagram

moreno-moises committed:
fix: removed 'action' json element + new load file variables

actions are now defined through an elements id.

variables use when loading a json has been renamed in order to correctly represent the loading process

moreno-moises committed:
add: doxygen style commenting in h files

moreno-moises committed:
add: doxygen style commenting in h file

moreno-moises committed:
add: error logging outputs to console

moreno-moises committed:
Merge branch 'feat-v0' of https://github.com/moreno-moises/ofxGuiJson into feat-v0

moreno-moises committed:
add future works to readme

moreno-moises committed:
Merge branch 'main' into feat-v0
